File: internal/rule/rule.go
package rule

import (
	"regexp"
	"strings"
)

type Action string

const (
	ActionPassthrough Action = "passthrough"
	ActionModify      Action = "modify"
	ActionBlock       Action = "block"
	ActionRedirect    Action = "redirect"
)

type MatchField string

const (
	FieldHost   MatchField = "host"
	FieldPath   MatchField = "path"
	FieldMethod MatchField = "method"
	FieldHeader MatchField = "header"
	FieldBody   MatchField = "body"
)

// Condition is a single match condition.
type Condition struct {
	Field   MatchField `json:"field"`
	Op      string     `json:"op"`    // contains, equals, regex, prefix
	Value   string     `json:"value"`
	Negate  bool       `json:"negate,omitempty"`
	compiled *regexp.Regexp
}

// Modification describes a header/body change to apply.
type Modification struct {
	Target    string `json:"target"`    // req_header, resp_header, req_body, resp_body
	Key       string `json:"key,omitempty"`
	Value     string `json:"value,omitempty"`
	Operation string `json:"operation"` // set, delete, replace, append
	Find      string `json:"find,omitempty"`
	Replace   string `json:"replace,omitempty"`
}

// Rule is an intercept rule with conditions and actions.
type Rule struct {
	ID           int64          `json:"id"`
	Name         string         `json:"name"`
	Enabled      bool           `json:"enabled"`
	Priority     int            `json:"priority"`
	Conditions   []Condition    `json:"conditions"`
	Action       Action         `json:"action"`
	Modifications []Modification `json:"modifications,omitempty"`
	RedirectURL  string         `json:"redirect_url,omitempty"`
}

// Manager holds all rules and applies them.
type Manager struct {
	rules []*Rule
}

func NewManager() *Manager {
	return &Manager{}
}

func (m *Manager) Add(r *Rule) {
	m.rules = append(m.rules, r)
	m.sort()
}

func (m *Manager) Update(r *Rule) {
	for i, existing := range m.rules {
		if existing.ID == r.ID {
			m.rules[i] = r
			m.sort()
			return
		}
	}
}

func (m *Manager) Delete(id int64) {
	for i, r := range m.rules {
		if r.ID == id {
			m.rules = append(m.rules[:i], m.rules[i+1:]...)
			return
		}
	}
}

func (m *Manager) List() []*Rule {
	out := make([]*Rule, len(m.rules))
	copy(out, m.rules)
	return out
}

// Match returns the first matching enabled rule, or nil.
func (m *Manager) Match(ctx *MatchContext) *Rule {
	for _, r := range m.rules {
		if !r.Enabled {
			continue
		}
		if matchAll(r.Conditions, ctx) {
			return r
		}
	}
	return nil
}

func (m *Manager) sort() {
	for i := 1; i < len(m.rules); i++ {
		for j := i; j > 0 && m.rules[j].Priority > m.rules[j-1].Priority; j-- {
			m.rules[j], m.rules[j-1] = m.rules[j-1], m.rules[j]
		}
	}
}

// MatchContext carries the data used to evaluate conditions.
type MatchContext struct {
	Method  string
	Host    string
	Path    string
	Headers map[string][]string
	Body    []byte
}

func matchAll(conds []Condition, ctx *MatchContext) bool {
	for _, c := range conds {
		if !matchOne(c, ctx) {
			return false
		}
	}
	return true
}

func matchOne(c Condition, ctx *MatchContext) bool {
	var val string
	switch c.Field {
	case FieldHost:
		val = ctx.Host
	case FieldPath:
		val = ctx.Path
	case FieldMethod:
		val = ctx.Method
	case FieldHeader:
		for k, vs := range ctx.Headers {
			if strings.EqualFold(k, c.Key()) {
				val = strings.Join(vs, ", ")
				break
			}
		}
	case FieldBody:
		val = string(ctx.Body)
	}

	var matched bool
	switch c.Op {
	case "equals":
		matched = val == c.Value
	case "contains":
		matched = strings.Contains(val, c.Value)
	case "prefix":
		matched = strings.HasPrefix(val, c.Value)
	case "regex":
		if c.compiled == nil {
			c.compiled, _ = regexp.Compile(c.Value)
		}
		if c.compiled != nil {
			matched = c.compiled.MatchString(val)
		}
	default:
		matched = strings.Contains(val, c.Value)
	}

	if c.Negate {
		return !matched
	}
	return matched
}

func (c *Condition) Key() string {
	// For header field, Value may be "Header-Name: pattern" format
	if idx := strings.Index(c.Value, ":"); idx > 0 {
		return strings.TrimSpace(c.Value[:idx])
	}
	return c.Value
}

File: paxy/rule/rule.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import StrEnum


class Action(StrEnum):
    PASSTHROUGH = "passthrough"
    MODIFY = "modify"
    BLOCK = "block"
    REDIRECT = "redirect"


class MatchField(StrEnum):
    HOST = "host"
    PATH = "path"
    METHOD = "method"
    HEADER = "header"
    BODY = "body"


@dataclass
class Condition:
    field: MatchField
    op: str  # equals, contains, prefix, regex
    value: str
    negate: bool = False
    _compiled: re.Pattern | None = field(default=None, init=False, repr=False)

    def matches(self, ctx: MatchContext) -> bool:
        val = self._extract(ctx)
        result = self._apply_op(val)
        return not result if self.negate else result

    def _extract(self, ctx: MatchContext) -> str:
        if self.field == MatchField.HOST:
            return ctx.host
        if self.field == MatchField.PATH:
            return ctx.path
        if self.field == MatchField.METHOD:
            return ctx.method
        if self.field == MatchField.BODY:
            return ctx.body.decode(errors="replace")
        if self.field == MatchField.HEADER:
            name = (
                self.value.split(":")[0].strip().lower()
                if ":" in self.value
                else self.value.lower()
            )
            for k, vs in ctx.headers.items():
                if k.lower() == name:
                    return ", ".join(vs)
        return ""

    def _apply_op(self, val: str) -> bool:
        if self.op == "equals":
            return val == self.value
        if self.op == "contains":
            return self.value in val
        if self.op == "prefix":
            return val.startswith(self.value)
        if self.op == "regex":
            if self._compiled is None:
                self._compiled = re.compile(self.value)
            return bool(self._compiled.search(val))
        return self.value in val


@dataclass
class Modification:
    target: str  # req_header, resp_header, req_body, resp_body
    operation: str  # set, delete, append, replace, find_replace
    key: str = ""
    value: str = ""
    find: str = ""
    replace: str = ""


@dataclass
class Rule:
    id: int = 0
    name: str = ""
    enabled: bool = True
    priority: int = 0
    conditions: list[Condition] = field(default_factory=list)
    action: Action = Action.PASSTHROUGH
    modifications: list[Modification] = field(default_factory=list)
    redirect_url: str = ""

    def matches(self, ctx: MatchContext) -> bool:
        return all(c.matches(ctx) for c in self.conditions)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "enabled": self.enabled,
            "priority": self.priority,
            "conditions": [
                {
                    "field": c.field.value,
                    "op": c.op,
                    "value": c.value,
                    "negate": c.negate,
                }
                for c in self.conditions
            ],
            "action": self.action.value,
            "modifications": [
                {
                    "target": m.target,
                    "operation": m.operation,
                    "key": m.key,
                    "value": m.value,
                    "find": m.find,
                    "replace": m.replace,
                }
                for m in self.modifications
            ],
            "redirect_url": self.redirect_url,
        }

    @classmethod
    def from_dict(cls, data: dict) -> Rule:
        rule = cls(
            id=data.get("id", 0),
            name=data.get("name", ""),
            enabled=data.get("enabled", True),
            priority=data.get("priority", 0),
            action=Action(data.get("action", "passthrough")),
            redirect_url=data.get("redirect_url", ""),
        )
        for c in data.get("conditions", []):
            rule.conditions.append(
                Condition(
                    field=MatchField(c["field"]),
                    op=c.get("op", "contains"),
                    value=c.get("value", ""),
                    negate=c.get("negate", False),
                )
            )
        for m in data.get("modifications", []):
            rule.modifications.append(
                Modification(
                    target=m["target"],
                    operation=m.get("operation", "set"),
                    key=m.get("key", ""),
                    value=m.get("value", ""),
                    find=m.get("find", ""),
                    replace=m.get("replace", ""),
                )
            )
        return rule


@dataclass
class MatchContext:
    method: str
    host: str
    path: str
    headers: dict[str, list[str]]
    body: bytes


class RuleManager:
    def __init__(self) -> None:
        self._rules: list[Rule] = []
        self._counter = 0

    def add(self, rule: Rule) -> Rule:
        self._counter += 1
        rule.id = self._counter
        self._rules.append(rule)
        self._sort()
        return rule

    def update(self, rule: Rule) -> None:
        for i, r in enumerate(self._rules):
            if r.id == rule.id:
                self._rules[i] = rule
                self._sort()
                return

    def delete(self, rule_id: int) -> None:
        self._rules = [r for r in self._rules if r.id != rule_id]

    def list(self) -> list[Rule]:
        return list(self._rules)

    def match(self, ctx: MatchContext) -> Rule | None:
        for rule in self._rules:
            if rule.enabled and rule.matches(ctx):
                return rule
        return None

    def _sort(self) -> None:
        self._rules.sort(key=lambda r: r.priority, reverse=True)
